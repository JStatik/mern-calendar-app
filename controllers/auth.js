const { response } = require( 'express' );
const bcrypt = require( 'bcryptjs' );
const Usuario = require( '../models/Usuario' );
const { generarJWT } = require( '../helpers/jwt' );

const loginUsuario = async( req, res = response ) => {
    const { email: emailReq, password: passwordReq } = req.body;

    try {
        const usuario = await Usuario.findOne( { email: emailReq } );

        if( !usuario ) {
            return res.status( 400 ).json(
                {
                    ok: false,
                    msg: 'Los datos ingresados no son válidos'
                }
            );
        }

        const validPassword = bcrypt.compareSync( passwordReq, usuario.password );

        if( !validPassword ) {
            return res.status( 400 ).json(
                {
                    ok: false,
                    msg: 'Los datos ingresados no son válidos'
                }
            );
        }

        const token = await generarJWT( usuario.id, usuario.name );

        return res.status( 200 ).json(
            {
                ok: true,
                uid: usuario.id,
                name: usuario.name,
                token: token
            }
        );
    } catch( error ) {
        console.log( error );

        return res.status( 500 ).json(
            {
                ok: false,
                msg: 'Por favor, hable con el administrador'
            }
        );
    }
};

const crearUsuario = async( req, res = response ) => {
    const { email: emailReq, password: passwordReq } = req.body;

    try {
        let usuario = await Usuario.findOne( { email: emailReq } );

        if( usuario ) {
            return res.status( 400 ).json(
                {
                    ok: false,
                    msg: 'El email ingresado ya está en uso'
                }
            );
        }

        usuario = new Usuario( req.body );

        const salt = bcrypt.genSaltSync();
        usuario.password = bcrypt.hashSync( passwordReq, salt );

        await usuario.save();

        const token = await generarJWT( usuario.id, usuario.name );

        return res.status( 201 ).json(
            {
                ok: true,
                uid: usuario.id,
                name: usuario.name,
                token: token
            }
        );
    } catch( error ) {
        console.log( error );

        return res.status( 500 ).json(
            {
                ok: false,
                msg: 'Por favor, hable con el administrador'
            }
        );
    }
};

const revalidarToken = async( req, res = response ) => {
    const { uid, name } = req;

    const token = await generarJWT( uid, name );
    
    return res.json(
        {
            ok: true,
            uid: uid,
            name: name,
            token: token
        }
    );
};

module.exports = {
    loginUsuario,
    crearUsuario,
    revalidarToken
};
