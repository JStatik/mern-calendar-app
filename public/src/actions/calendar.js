import { fetchEvents } from "../helpers/fetch";
import popUp from "../helpers/popup";
import prepareEvents from "../helpers/prepareEvents";
import types from "../types/types";

export const startLoadEvents = () => {
    return async( dispatch ) => {
        const token = localStorage.getItem( 'tkca' ) || '';

        try {
            const peticion = await fetchEvents( '/events', 'GET', token );
            const eventos = await peticion.json();

            const eventosPreparados = prepareEvents( eventos.eventos );

            dispatch( loadEvents( eventosPreparados ) );
        } catch( error ) {
            popUp( 'error', '#990000', 'Error', 'No se pudo obtener los eventos, intente nuevamente.' );
        }
    };
};

export const startDeleteEvent = () => {
    return async( dispatch, getState ) => {
        const { idEvent } = getState().calendar.activeEvent;
        const token = localStorage.getItem( 'tkca' ) || '';

        try {
            const peticion = await fetchEvents( `/events/${ idEvent }`, 'DELETE', token );
            const evento = await peticion.json();

            if( evento.ok ) {
                dispatch( deleteEvent() );
            } else {
                dispatch( removeActiveEvent() );
                popUp( 'error', '#990000', 'Error', 'No tiene privilegios de eliminación sobre este evento' );
            }
        } catch( error ) {
            popUp( 'error', '#990000', 'Error', 'No se pudo eliminar el evento, intente nuevamente.' );
        }
    };
};

export const startUpdateEvent = ( event ) => {
    return async( dispatch, getState ) => {
        const { idEvent } = getState().calendar.activeEvent;
        const token = localStorage.getItem( 'tkca' ) || '';

        try {
            const peticion = await fetchEvents( `/events/${ idEvent }`, 'PUT', token, event );
            const evento = await peticion.json();

            if( evento.ok ) {
                event.user = evento.evento.user;
                event.idEvent = evento.evento.id;

                dispatch( updateEvent( event ) );
            } else {
                popUp( 'error', '#990000', 'Error', 'No tiene privilegios de edición sobre este evento' );
            }
        } catch( error ) {
            popUp( 'error', '#990000', 'Error', 'No se pudo actualizar el evento, intente nuevamente.' );
        }
    };
};

export const startAddNewEvent = ( event ) => {
    return async( dispatch ) => {
        const token = localStorage.getItem( 'tkca' ) || '';

        try {
            const peticion = await fetchEvents( '/events', 'POST', token, event );
            const evento = await peticion.json();

            if( evento.ok ) {
                event.user = evento.evento.user;
                event.idEvent = evento.evento.id;

                dispatch( addNewEvent( event ) );
            }
        } catch( error ) {
            popUp( 'error', '#990000', 'Error', 'No se pudo crear el evento, intente nuevamente.' );
        }
    };
};

const loadEvents = ( events ) => (
    {
        type: types.eventLoad,
        payload: events
    }
);

const deleteEvent = () => (
    {
        type: types.eventDelete
    }
);

const updateEvent = ( event ) => (
    {
        type: types.eventUpdate,
        payload: event
    }
);

const addNewEvent = ( event ) => (
    {
        type: types.eventAddNew,
        payload: event
    }
);

export const removeActiveEvent = () => (
    {
        type: types.eventRemoveActive
    }
);

export const setActiveEvent = ( event ) => (
    {
        type: types.eventSetActive,
        payload: event
    }
);

export const removeDatesSlot = () => (
    {
        type: types.eventRemoveDateSlot
    }
);

export const setDatesSlot = ( startSlot, endSlot ) => (
    {
        type: types.eventSetDateSlot,
        payload: {
            start: startSlot,
            end: endSlot
        }
    }
);

export const removeEventsLogout = () => (
    {
        type: types.eventRemove
    }
);
