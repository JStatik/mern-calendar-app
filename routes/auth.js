/*********************************************************************** RUTAS DE USUARIO / AUTH ***********************************************************************/
/*************************************************************************** HOST + /API/AUTH ***************************************************************************/
const { Router } = require( 'express' );
const { check } = require( 'express-validator' );
const { loginUsuario, crearUsuario, revalidarToken } = require( '../controllers/auth' );
const { validarCampos } = require( '../middlewares/validar_campos' );
const { validarJWT } = require('../middlewares/varlidar_jwt');

const router = Router();

router.post( 
    '/',
    [
        check( 'email', 'Debe ingresar un email válido' ).isEmail().normalizeEmail(),
        check( 'password', 'Debe ingresar un password válido: Al menos 6 caracteres y alfanúmerico' ).isLength( { min: 6 } ).isAlphanumeric(),
        validarCampos
    ],
    loginUsuario
);

router.post( 
    '/register', 
    [ 
        check( 'name', 'Debe ingresar un nombre válido' ).trim().escape().not().isEmpty().isLength( { min: 2 } ).isString(),
        check( 'email', 'Debe ingresar un email válido' ).isEmail().normalizeEmail(),
        check( 'password', 'Debe ingresar un password válido: Al menos 6 caracteres y alfanúmerico' ).isLength( { min: 6 } ).isAlphanumeric(),
        validarCampos
    ], 
    crearUsuario 
);

router.get( '/renew', validarJWT, revalidarToken );

module.exports = router;
