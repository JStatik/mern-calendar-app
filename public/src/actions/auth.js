import types from '../types/types';
import { fetchLoginAndRegister, fetchRenew } from '../helpers/fetch';
import popUp from '../helpers/popup';
import { removeEventsLogout } from './calendar';

export const startChecking = () => {
    return async( dispatch ) => {
        const token = localStorage.getItem( 'tkca' ) || '';

        try {
            const peticion = await fetchRenew( '/auth/renew', token );
            const usuario = await peticion.json();

            if( usuario.ok ) {
                localStorage.setItem( 'tkca', usuario.token );

                dispatch( login( {
                    uid: usuario.uid,
                    name: usuario.name
                } ) );
            } else {
                dispatch( checkingFinish() );
                dispatch( startLogout() );
            }
        } catch( error ) {
            localStorage.removeItem( 'tkca' );
        }
    };
};

export const startRegister = ( name, email, password ) => {
    return async( dispatch ) => {
        const data = {
            name: name,
            email: email,
            password: password
        };

        try {
            const peticion = await fetchLoginAndRegister( '/auth/register', data );
            const usuario = await peticion.json();

            if( usuario.ok ) {
                localStorage.setItem( 'tkca', usuario.token );

                dispatch( login( {
                    uid: usuario.uid,
                    name: usuario.name
                } ) );
            } else {
                popUp( 'error', '#990000', 'Error', 'El email ingresado ya está en uso.' );
            }
        } catch( error ) {
            popUp( 'error', '#990000', 'Error', 'Error al registrarse, intente nuevamente.' );
        }
    };
};

export const startLogin = ( email, password ) => {
    return async( dispatch ) => {
        const data = {
            email: email,
            password: password
        };

        try {
            const peticion = await fetchLoginAndRegister( '/auth', data );
            const usuario = await peticion.json();

            if( usuario.ok ) {
                localStorage.setItem( 'tkca', usuario.token );

                dispatch( login( {
                    uid: usuario.uid,
                    name: usuario.name
                } ) );
            } else {
                popUp( 'error', '#990000', 'Error', 'Alguno de los datos ingresados no es correcto.' );
            }
        } catch( error ) {
            popUp( 'error', '#990000', 'Error', 'Error al iniciar sesión, intente nuevamente.' );
        }
    };
};

export const startLogout = () => {
    return ( dispatch ) => {
        localStorage.removeItem( 'tkca' );

        dispatch( removeEventsLogout() );              
        dispatch( logout() );
    };
};

export const checkingFinish = () => (
    {
        type: types.authCheckingFinish
    }
);

const login = ( user ) => (
    {
        type: types.authLogin,
        payload: user
    }
);

const logout = () => (
    {
        type: types.authLogout
    }
);
