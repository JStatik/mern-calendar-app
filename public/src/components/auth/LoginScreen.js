import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import useForm from '../../hooks/useForm';
import { isFormValidLogin, isFormValidRegister } from '../../helpers/isFormValid';
import { startLogin, startRegister } from '../../actions/auth';
import './login.css';

const LoginScreen = () => {
    const dispatch = useDispatch();

    const [ disabled, setDisabled ] = useState( false );
    const [ msgErrorLogin, setMsgErrorLogin ] = useState( null );
    const [ msgErrorRegisterName, setMsgErrorRegisterName ] = useState( null );
    const [ msgErrorRegisterEmail, setMsgErrorRegisterEmail ] = useState( null );
    const [ msgErrorRegisterPassword, setMsgErrorRegisterPassword ] = useState( null );

    const [ formLoginValues, setFormLoginValues, handleLoginInputChange, resetLogin ] = useForm( { loginEmail: '', loginPassword: '' } );
    const { loginEmail, loginPassword } = formLoginValues;

    const [ formRegisterValues, setFormRegisterValues, handleRegisterInputChange, resetRegister ] = useForm( { registerName: '', registerEmail: '', registerPassword: '', registerPassword2: '' } );
    const { registerName, registerEmail, registerPassword, registerPassword2 } = formRegisterValues;

    const handleLoginSubmit = ( event ) => {
        event.preventDefault();

        const { newMsgError, isValid } = isFormValidLogin( loginEmail, loginPassword );
        setMsgErrorLogin( newMsgError );

        if( isValid ) {
            setDisabled( true );

            dispatch( startLogin( loginEmail, loginPassword ) );

            setDisabled( false );
        }

        resetLogin();
    };

    const handleRegisterSubmit = ( event ) => {
        event.preventDefault();

        const { newMsgErrorName, newMsgErrorEmail, newMsgErrorPassword, isValid } = isFormValidRegister( registerName, registerEmail, registerPassword, registerPassword2 );
        setMsgErrorRegisterName( newMsgErrorName );
        setMsgErrorRegisterEmail( newMsgErrorEmail );
        setMsgErrorRegisterPassword( newMsgErrorPassword );

        if( isValid ) {
            setDisabled( true );

            dispatch( startRegister( registerName, registerEmail, registerPassword ) );

            setDisabled( false );

            resetRegister();
        }
    };

    const handleClickSignUp = () => {
        setDisabled( true );

        document.getElementById( 'container' ).classList.add( 'right-panel-active' );
        setFormLoginValues( { loginEmail: '', loginPassword: '' } );
        setMsgErrorLogin( null );

        setDisabled( false );
    };

    const handleClickSignIn = () => {
        setDisabled( true );

        document.getElementById( 'container' ).classList.remove( 'right-panel-active' );
        setFormRegisterValues( { registerName: '', registerEmail: '', registerPassword: '', registerPassword2: '' } );
        setMsgErrorRegisterName( null );
        setMsgErrorRegisterEmail( null );
        setMsgErrorRegisterPassword( null );

        setDisabled( false );
    };

    return (
        <div className="form-register">
            <div className="container-login animate__animated animate__fadeIn" id="container">
                <div className="form-container sign-up-container">
                    <form className="form" autoComplete="off" onSubmit={ handleRegisterSubmit }>
                        <h3>Register</h3>
                        { msgErrorRegisterName && <div className="alert-error">{ msgErrorRegisterName }</div> }
                        { msgErrorRegisterEmail && <div className="alert-error">{ msgErrorRegisterEmail }</div> }
                        { msgErrorRegisterPassword && <div className="alert-error">{ msgErrorRegisterPassword }</div> }
                        <input type="text" placeholder="Nombre" className={ `form-control input-login ${ msgErrorRegisterName && 'is-invalid' }` } name="registerName" value={ registerName } onChange={ handleRegisterInputChange }/>
                        <input type="email" placeholder="Email" className={ `form-control input-login ${ msgErrorRegisterEmail && 'is-invalid' }` } name="registerEmail" value={ registerEmail } onChange={ handleRegisterInputChange }/>
                        <input type="password" placeholder="Contraseña" className={ `form-control input-login ${ msgErrorRegisterPassword && 'is-invalid' }` } name="registerPassword" value={ registerPassword } onChange={ handleRegisterInputChange }/>
                        <input type="password" placeholder="Repetir contraseña" className={ `form-control input-login ${ msgErrorRegisterPassword && 'is-invalid' }` } name="registerPassword2" value={ registerPassword2 } onChange={ handleRegisterInputChange }/>
                        <button className="btn-form" disabled={ disabled }>Crear cuenta</button>
                    </form>
                </div>

                <div className="form-container sign-in-container">
                    <form className="form" autoComplete="off" onSubmit={ handleLoginSubmit }>
                        <h3>Login</h3>
                        { msgErrorLogin && <div className="alert-error">{ msgErrorLogin }</div> }
                        <input type="email" placeholder="Email" className={ `form-control input-login ${ msgErrorLogin && 'is-invalid' }` } name="loginEmail" value={ loginEmail } onChange={ handleLoginInputChange }/>
                        <input type="password" placeholder="Contraseña" className={ `form-control input-login ${ msgErrorLogin && 'is-invalid' }` } name="loginPassword" value={ loginPassword } onChange={ handleLoginInputChange }/>
                        <button className="btn-form" disabled={ disabled }>Ingresar</button>
                    </form>
                </div>

                <div className="overlay-container">
                    <div className="overlay">
                        <div className="overlay-panel overlay-left">
                            <h4>¡Bienvenido!</h4>
                            <p>Para mantenerse conectado con nosotros, inicie sesión</p>
                            <button className="btn-form ghost" id="signIn" onClick={ handleClickSignIn } disabled={ disabled }>Login</button>
                        </div>

                        <div className="overlay-panel overlay-right">
                            <h4>¡Hola!</h4>
                            <p>Ingrese sus datos y comience su viaje con nosotros</p>
                            <button className="btn-form ghost" id="signUp" onClick={ handleClickSignUp } disabled={ disabled }>Register</button>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    )
};

export default LoginScreen;
