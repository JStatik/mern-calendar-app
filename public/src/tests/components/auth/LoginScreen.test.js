import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import LoginScreen from '../../../components/auth/LoginScreen';
import { startLogin, startRegister } from '../../../actions/auth';
jest.mock( '../../../actions/auth', () => (
    {
        startLogin: jest.fn(),
        startRegister: jest.fn()
    }
) );

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const store = mockStore( {} );
store.dispatch = jest.fn();

describe( 'Pruebas en LoginScreen', () => {
    const wrapper = mount( 
        <Provider store={ store }>
            <LoginScreen />
        </Provider>
    );

    beforeEach( () => {
        store.clearActions();
        jest.clearAllMocks();
    } );

    test( 'Debe mostrar el componente correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    } );

    test( 'Debe ejecutar el dispatch startLogin()', () => {
        wrapper.find( 'input[name="loginEmail"]' ).simulate( 'change', {
            target: {
                name: 'loginEmail',
                value: 'pepe@gmail.com'
            }
        } );

        wrapper.find( 'input[name="loginPassword"]' ).simulate( 'change', {
            target: {
                name: 'loginPassword',
                value: '123456'
            }
        } );

        wrapper.find( 'form' ).at( 1 ).simulate( 'submit', {
            preventDefault(){}
        } );

        expect( store.dispatch ).toHaveBeenCalled();
        expect( startLogin ).toHaveBeenCalledWith( 'pepe@gmail.com', '123456' );
    } );

    test( 'No debe ejecutar el dispatch startRegister() por contraseÃ±as diferentes', () => {
        wrapper.find( 'input[name="registerName"]' ).simulate( 'change', {
            target: {
                name: 'registerName',
                value: 'Pepe Ruiz'
            }
        } );

        wrapper.find( 'input[name="registerEmail"]' ).simulate( 'change', {
            target: {
                name: 'registerEmail',
                value: 'pepe@gmail.com'
            }
        } );

        wrapper.find( 'input[name="registerPassword"]' ).simulate( 'change', {
            target: {
                name: 'registerPassword',
                value: '123456'
            }
        } );

        wrapper.find( 'input[name="registerPassword2"]' ).simulate( 'change', {
            target: {
                name: 'registerPassword2',
                value: '1234567'
            }
        } );

        wrapper.find( 'form' ).at( 0 ).simulate( 'submit', {
            preventDefault(){}
        } );

        expect( store.dispatch ).not.toHaveBeenCalled();
        expect( startRegister ).not.toHaveBeenCalled();
    } );
} );
