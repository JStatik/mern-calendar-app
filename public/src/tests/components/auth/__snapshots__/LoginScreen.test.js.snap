// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en LoginScreen Debe mostrar el componente correctamente 1`] = `
<div
  className="form-register"
>
  <div
    className="container-login animate__animated animate__fadeIn"
    id="container"
  >
    <div
      className="form-container sign-up-container"
    >
      <form
        autoComplete="off"
        className="form"
        onSubmit={[Function]}
      >
        <h3>
          Register
        </h3>
        <input
          className="form-control input-login null"
          name="registerName"
          onChange={[Function]}
          placeholder="Nombre"
          type="text"
          value=""
        />
        <input
          className="form-control input-login null"
          name="registerEmail"
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value=""
        />
        <input
          className="form-control input-login null"
          name="registerPassword"
          onChange={[Function]}
          placeholder="Contraseña"
          type="password"
          value=""
        />
        <input
          className="form-control input-login null"
          name="registerPassword2"
          onChange={[Function]}
          placeholder="Repetir contraseña"
          type="password"
          value=""
        />
        <button
          className="btn-form"
          disabled={false}
        >
          Crear cuenta
        </button>
      </form>
    </div>
    <div
      className="form-container sign-in-container"
    >
      <form
        autoComplete="off"
        className="form"
        onSubmit={[Function]}
      >
        <h3>
          Login
        </h3>
        <input
          className="form-control input-login null"
          name="loginEmail"
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value=""
        />
        <input
          className="form-control input-login null"
          name="loginPassword"
          onChange={[Function]}
          placeholder="Contraseña"
          type="password"
          value=""
        />
        <button
          className="btn-form"
          disabled={false}
        >
          Ingresar
        </button>
      </form>
    </div>
    <div
      className="overlay-container"
    >
      <div
        className="overlay"
      >
        <div
          className="overlay-panel overlay-left"
        >
          <h4>
            ¡Bienvenido!
          </h4>
          <p>
            Para mantenerse conectado con nosotros, inicie sesión
          </p>
          <button
            className="btn-form ghost"
            disabled={false}
            id="signIn"
            onClick={[Function]}
          >
            Login
          </button>
        </div>
        <div
          className="overlay-panel overlay-right"
        >
          <h4>
            ¡Hola!
          </h4>
          <p>
            Ingrese sus datos y comience su viaje con nosotros
          </p>
          <button
            className="btn-form ghost"
            disabled={false}
            id="signUp"
            onClick={[Function]}
          >
            Register
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
